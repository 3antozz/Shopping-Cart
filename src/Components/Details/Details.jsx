import styles from "./Details.module.css";
import PropTypes from "prop-types";
import {useLocation, useOutletContext } from "react-router-dom";
import { Star } from 'lucide-react';
import { useState } from "react";

export default function Details () {
    const { state } = useLocation();
    const { handleAddToCart } = useOutletContext();
    const [inputValue, setInputValue] = useState(1);
    const [popup, setPopup] = useState(false);

    function handlePopup () {
        setPopup(true);
        setTimeout(() => {
            setPopup(false)
        }, 4000)
    }

    function incrementInput (event) {
        setInputValue((prev) => {
            if (event.target.className === "increment") {
                return +prev + 1;
            } else {
                if (prev > 1) {
                    return +prev - 1;
                }
                return +prev;
            }
        })
    }
    function handleInput (event) {
        setInputValue(event.target.value);
    }

    return (
        <div className={styles.detail}>
            <div className={styles.details}>
            {popup && <div className={styles.popup}>Item added to the cart!</div>}
            <img src={state.image} alt={state.title} />
            <div className={styles.right}>
                <h1>{state.title}</h1>
                <h2 className={styles.rating}><Rating rating={state.rating.rate}/> ({state.rating.count})</h2>
                <h2>Price: {state.price}$</h2>
                <h2>Description:</h2>
                <p>{state.description}</p>
                <form className={styles.quantity} onSubmit={(event) => handleAddToCart(event, state.id, +inputValue)}>
                    <div>
                        <label htmlFor="quantity"></label>
                        <button type="button" onClick={incrementInput}>-</button>
                        <input type="number" id="quantity" min={1} max={100} value={inputValue} onChange={handleInput}  />
                        <button type="button" onClick={incrementInput} className="increment">+</button>
                    </div>
                    <button type="submit" onClick={handlePopup}>Add to Cart</button>
                </form>
            </div>
        </div>
        </div>
    )

}

const Rating = ({rating}) => {
    const rate = Math.round(rating);
    let array = [];
    for (let i=0; i < 5; i++) {
        i < rate ? array.push(true) : array.push(false)
    }

    return (
        <div className={styles.star}>
            {array.map((bool, index) => <Star key={index} size={30} color='grey' fill={bool ? "yellow" : "#ffffff00"} />)}
        </div>
    )
}

Rating.propTypes = {
    rating: PropTypes.number.isRequired,
};