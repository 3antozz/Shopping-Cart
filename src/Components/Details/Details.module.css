.detail {
    flex: 1;
    padding: 2rem 10%;
    align-content: center;
}

.details {
    display: flex;
    align-items: center;
    padding: 3%;
    gap: 10%;
    box-shadow: 1px 1px 4px 1px #80808061;
    border-radius: 20px;
}

.details img {
    min-width: 330px;
    width: 330px;
    height: 350px;
    object-fit: contain;
    flex-shrink: 0;
}

.details h1 {
    font-size: clamp(1.5rem, 2.7vw, 2rem);
}

.details h2 {
    font-size: clamp(1.2rem, 1.8vw, 1.35rem);
}

.details p {
    font-size: clamp(1rem, 1.2vw, 1.3rem);
}

.rating {
    display: flex;
    align-items: center;
    gap: 1rem;
}

.star {
    display: flex;
}

.right > * {
    margin-bottom: 0.7rem;
}


.rating {
    display: flex;
}

.quantity {
    display: flex;
    align-items: center;
    gap: 2rem;
}

.quantity label {
    display: none;
}

.quantity input {
    width: 3.5rem;
    padding: 5px 5px;
    border-radius: 5px;
    border: 1px solid black;
    text-align: center;
}

.quantity button[type="button"] {
    border-radius: 10px;
    background: none;
    border: none;
    width: 30px;
    font-size: 1.5rem;
    color: black;
}

.quantity button[type="button"]:hover {
    background: red;
    color: white;
}

.quantity button[type="button"]:active {
    transform: scale(1.1);
}

.quantity button[type="submit"] {
    font-weight: bold;
    border-radius: 15px;
    font-size: 1.4rem;
    padding: 0.4rem 1.2rem;
    background-color: rgba(255, 0, 0, 0.745);
    color: white;
    border: 1px solid black;
    transition: transform 200ms ease;
}

.quantity button[type="submit"]:hover {
    transform: scale(1.1);
    background-color: rgb(210, 3, 3);
    cursor: pointer;
}

.quantity > div {
    display: flex;
    gap: 0.5rem;
}

@media (max-width: 768px) {
    .details {
        flex-direction: column;
        text-align: center;
        gap: 2.5rem;
    }

    .rating {
        justify-content: center;
    }

    .details img {
        min-width: 150px;
        width: 50%;
    }
    
}