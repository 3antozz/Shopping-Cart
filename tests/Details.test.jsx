import { describe, it, expect, beforeEach, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Details from "../src/Components/Details/Details.jsx";
import { MemoryRouter, Outlet, Routes, Route } from "react-router-dom";

describe("Details Component", () => {
    const handleAddToCart = vi.fn()
    const context = {
        loading: false,
        handleAddToCart,
        products: [{
                id: 1,
                title: "test",
                price: 69,
                category: "...",
                description: "...",
                image: "...",
                rating: {
                    rate: 102,
                    count: 69
                }
            }, {
                id: 2,
                title: "test2",
                price: 69,
                category: "...",
                description: "...",
                image: "...",
                rating: {
                    rate: 102,
                    count: 69
                }
            }]
    }
    beforeEach(() => {
        render(
            <MemoryRouter initialEntries={["/shop/products/1"]}>
            <Routes>
              <Route path="/shop" element={<div>
                <h1>SHOP ROUTE</h1>
                <Outlet context={context} />
              </div>}>
                <Route path="products/:productId" element={<Details />} />
              </Route>
            </Routes>
          </MemoryRouter>
        );

    })
    it("should render component", () => {
        expect(screen.getByTestId("container")).toBeInTheDocument();
    });

    it("calls functions on click", async() => {
        const user = userEvent.setup();
        const button = screen.getByRole("button", {name: /add to cart/i});
        await user.click(button);
        expect(handleAddToCart).toBeCalled();
        expect(screen.getByText(/Item added to the cart!/i)).toBeInTheDocument();
    })
    it("should increment and decrement when clicked", async () =>{
        const user = userEvent.setup();
        const incrementButton = screen.getByRole("button", {name: "+"});
        const decrementButton = screen.getByRole("button", {name: "-"});

        const input = screen.getByDisplayValue("1");
        await user.click(incrementButton);
        expect(input.value).toBe("2");
        await user.click(decrementButton);
        expect(input.value).toBe("1");
    })
        it("matches snapshot", () => {
            const handleAddToCart = vi.fn()
            const context = {
                loading: false,
                handleAddToCart,
                products: [{
                        id: 1,
                        title: "test",
                        price: 69,
                        category: "...",
                        description: "...",
                        image: "...",
                        rating: {
                            rate: 102,
                            count: 69
                        }
                    }, {
                        id: 2,
                        title: "test2",
                        price: 69,
                        category: "...",
                        description: "...",
                        image: "...",
                        rating: {
                            rate: 102,
                            count: 69
                        }
                    }]
            }
            const {container} = render(
                <MemoryRouter initialEntries={["/shop/products/1"]}>
                <Routes>
                  <Route path="/shop" element={<div>
                    <h1>SHOP ROUTE</h1>
                    <Outlet context={context} />
                  </div>}>
                    <Route path="products/:productId" element={<Details />} />
                  </Route>
                </Routes>
              </MemoryRouter>
            );
            expect(container).toMatchSnapshot();
        });
});