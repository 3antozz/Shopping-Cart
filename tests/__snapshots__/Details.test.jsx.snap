// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Details Component > matches snapshot 1`] = `
<div>
  <div>
    <h1>
      SHOP ROUTE
    </h1>
    <div
      class="_detail_f1ce5d"
      data-testid="container"
    >
      <div
        class="_details_f1ce5d"
      >
        <img
          alt="test"
          src="..."
        />
        <div
          class="_right_f1ce5d"
        >
          <h1>
            test
          </h1>
          <h2
            class="_rating_f1ce5d"
          >
            <div
              class="_star_f1ce5d"
            >
              <svg
                class="lucide lucide-star"
                fill="yellow"
                height="30"
                stroke="grey"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="30"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11.525 2.295a.53.53 0 0 1 .95 0l2.31 4.679a2.123 2.123 0 0 0 1.595 1.16l5.166.756a.53.53 0 0 1 .294.904l-3.736 3.638a2.123 2.123 0 0 0-.611 1.878l.882 5.14a.53.53 0 0 1-.771.56l-4.618-2.428a2.122 2.122 0 0 0-1.973 0L6.396 21.01a.53.53 0 0 1-.77-.56l.881-5.139a2.122 2.122 0 0 0-.611-1.879L2.16 9.795a.53.53 0 0 1 .294-.906l5.165-.755a2.122 2.122 0 0 0 1.597-1.16z"
                />
              </svg>
              <svg
                class="lucide lucide-star"
                fill="yellow"
                height="30"
                stroke="grey"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="30"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11.525 2.295a.53.53 0 0 1 .95 0l2.31 4.679a2.123 2.123 0 0 0 1.595 1.16l5.166.756a.53.53 0 0 1 .294.904l-3.736 3.638a2.123 2.123 0 0 0-.611 1.878l.882 5.14a.53.53 0 0 1-.771.56l-4.618-2.428a2.122 2.122 0 0 0-1.973 0L6.396 21.01a.53.53 0 0 1-.77-.56l.881-5.139a2.122 2.122 0 0 0-.611-1.879L2.16 9.795a.53.53 0 0 1 .294-.906l5.165-.755a2.122 2.122 0 0 0 1.597-1.16z"
                />
              </svg>
              <svg
                class="lucide lucide-star"
                fill="yellow"
                height="30"
                stroke="grey"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="30"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11.525 2.295a.53.53 0 0 1 .95 0l2.31 4.679a2.123 2.123 0 0 0 1.595 1.16l5.166.756a.53.53 0 0 1 .294.904l-3.736 3.638a2.123 2.123 0 0 0-.611 1.878l.882 5.14a.53.53 0 0 1-.771.56l-4.618-2.428a2.122 2.122 0 0 0-1.973 0L6.396 21.01a.53.53 0 0 1-.77-.56l.881-5.139a2.122 2.122 0 0 0-.611-1.879L2.16 9.795a.53.53 0 0 1 .294-.906l5.165-.755a2.122 2.122 0 0 0 1.597-1.16z"
                />
              </svg>
              <svg
                class="lucide lucide-star"
                fill="yellow"
                height="30"
                stroke="grey"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="30"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11.525 2.295a.53.53 0 0 1 .95 0l2.31 4.679a2.123 2.123 0 0 0 1.595 1.16l5.166.756a.53.53 0 0 1 .294.904l-3.736 3.638a2.123 2.123 0 0 0-.611 1.878l.882 5.14a.53.53 0 0 1-.771.56l-4.618-2.428a2.122 2.122 0 0 0-1.973 0L6.396 21.01a.53.53 0 0 1-.77-.56l.881-5.139a2.122 2.122 0 0 0-.611-1.879L2.16 9.795a.53.53 0 0 1 .294-.906l5.165-.755a2.122 2.122 0 0 0 1.597-1.16z"
                />
              </svg>
              <svg
                class="lucide lucide-star"
                fill="yellow"
                height="30"
                stroke="grey"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="30"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11.525 2.295a.53.53 0 0 1 .95 0l2.31 4.679a2.123 2.123 0 0 0 1.595 1.16l5.166.756a.53.53 0 0 1 .294.904l-3.736 3.638a2.123 2.123 0 0 0-.611 1.878l.882 5.14a.53.53 0 0 1-.771.56l-4.618-2.428a2.122 2.122 0 0 0-1.973 0L6.396 21.01a.53.53 0 0 1-.77-.56l.881-5.139a2.122 2.122 0 0 0-.611-1.879L2.16 9.795a.53.53 0 0 1 .294-.906l5.165-.755a2.122 2.122 0 0 0 1.597-1.16z"
                />
              </svg>
            </div>
             (
            69
            )
          </h2>
          <h2>
            Price: 
            69
            $
          </h2>
          <h2>
            Description:
          </h2>
          <p>
            ...
          </p>
          <form
            class="_quantity_f1ce5d"
          >
            <div>
              <label
                for="quantity"
              />
              <button
                type="button"
              >
                -
              </button>
              <input
                id="quantity"
                max="100"
                min="1"
                type="number"
                value="1"
              />
              <button
                class="increment"
                type="button"
              >
                +
              </button>
            </div>
            <button
              type="submit"
            >
              Add to Cart
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;
